new architecture to define all partitioned NLPModels

paraynaud opened the pull request:
@dpo, I'm working on the partitioned NLPModels.
This is a working version of the repository.
I will clean it later.
Now, PartiallySeparanleNLPModels define `PartiallySeparableNLPModel` and `PBFGSNLPModel, PSR1NLPModel,... <: AbstractPQNNLPModel`.

However, I still can't type correctly `PBFGSNLPModel` and the other partitioned quasi-Newton `NLPModels`.
It is strange because I typed it the same as I did for my previous version of `PartitionedData_TR_PQN` (used in the previous `PartiallySeparableNLPModel`).
This concerns the typing of all the compiled tapes of element function, I typed both `PartitionedData_TR_PQN` and `PBFGSNLPModel` with:
```julia
vec_compiled_element_gradients::Vector{ReverseDiff.CompiledTape}
```
and I use the same line to get allocate these tapes:
```julia
vec_compiled_element_gradients =
    map((tree -> compiled_grad_element_function(tree; type = T)::ReverseDiff.CompiledTape), element_expr_tree)
```
but I end up with an error of typing for `PBFGSNLPModel`.
```julia
ERROR: MethodError: no method matching PBFGSNLPModel(::ADNLPModel{Float64, Vector{Float64}}, ::NLPModelMeta{Float64, Vector{Float64}}, ::Counters, ::Int64, ::Int64, ::Vector{PartiallySeparableNLPModels.ModPSNLPModels.ElementFunction}, ::Int64, ::Vector{ExpressionTreeForge.M_implementation_tree.Type_node{ExpressionTreeForge.M_implementation_complete_expr_tree.Complete_node{Float64}}}, ::Vector{Vector{Int64}}, ::Vector{Int64}, ::Vector{ReverseDiff.CompiledTape{ReverseDiff.GradientTape{ExpressionTreeForge.var"#1#2"{ExpressionTreeForge.M_implementation_tree.Type_node{ExpressionTreeForge.M_abstract_expr_node.Abstract_expr_node}}, ReverseDiff.TrackedArray{Float64, Float64, 1, Vector{Float64}, Vector{Float64}}, ReverseDiff.TrackedReal{Float64, Float64, Nothing}}}}, ::Vector{Float64}, ::Vector{Float64}, ::Vector{Float64}, ::Elemental_pv{Float64}, ::Elemental_pv{Float64}, ::Elemental_pv{Float64}, ::Elemental_pv{Float64}, ::Elemental_pv{Float64}, ::Elemental_pm{Float64}, ::Float64, ::Symbol)
Closest candidates are:
  PBFGSNLPModel(::M, ::Meta, ::Counters, ::Int64, ::Int64, ::Vector{PartiallySeparableNLPModels.ModPSNLPModels.ElementFunction}, ::Int64, ::Vector{G}, ::Vector{Vector{Int64}}, ::Vector{Int64}, ::Vector{ReverseDiff.CompiledTape}, ::Vector{T}, ::Vector{T}, ::Vector{T}, ::Elemental_pv{T}, ::Elemental_pv{T}, ::Elemental_pv{T}, ::Elemental_pv{T}, ::Elemental_pv{T}, ::P, ::T, ::Symbol) where {G, P, T, S, M<:AbstractNLPModel{T, S}, Meta<:AbstractNLPModelMeta{T, S}} at c:\Users\raynaudp\Desktop\these\code\PartiallySeparableNLPModels.jl\src\new\partitionedNLPModels\PBFGSNLPModels.jl:12
  PBFGSNLPModel(::Union{NLPModelsJuMP.MathOptNLPModel, ADNLPModel}) at c:\Users\raynaudp\Desktop\these\code\PartiallySeparableNLPModels.jl\src\new\partitionedNLPModels\PBFGSNLPModels.jl:47
```
while its works fine for `PartitionedData_TR_PQN`.

The issue is that:
```julia
Vector{ReverseDiff.TrackedArray{Float64, Float64, 1, Vector{Float64}, Vector{Float64}}, ReverseDiff.TrackedReal{Float64, Float64, Nothing}}}}
```
is not reecognized as a 
```julia
Vector{ReverseDiff.CompiledTape}
```

I tried everything that came to my mind (especially copying what I did for the old `PartiallySeparableNLPModel`) and I run out of ideas now.

paraynaud commented:
Tests and documentaitons pass.

paraynaud commented:
@dpo, I reworked entirely the repository.
I will check tomorrow that I respect the NLPModels API.

paraynaud commented:
I implemented :
- `PBFGSNLPModel`;
- `PCSNLPModel`;
- `PLBFGSNLPModel`;
- `PLSR1NLPModel`;
- `PLSENLPModel`;
- `PSR1NLPModel`;
- `PSENLPModel`;
- `PSNLPModel`;
`<:AbstractPartiallySeparableNLPModels`, with a method `obj`, `grad!`, `hprod!`, `hessop` for each of them.

Do I need to define other methods?
maybe I can use (for ADNLPModels/nlp.jl):
```julia

# the following methods are not affected by the Hessian approximation
for meth in (
  :obj,
  :grad,
  :cons,
  :cons_lin,
  :cons_nln,
  :jac_coord,
  :jac_lin_coord,
  :jac_nln_coord,
  :jac,
  :jac_lin,
  :jac_nln,
)
  @eval NLPModels.$meth(nlp::QuasiNewtonModel, x::AbstractVector) = $meth(nlp.model, x)
```
?

dpo commented on docs/src/tutorial.md:
```diff
@@ -65,15 +70,27 @@ function jump_example(n::Int)
+There are several variants:
+* 'PBFGSNLPModel': every element-Hessian approximation is updated with BFGS;
+* 'PSR1NLPModel': every element-Hessian approximation is updated with SR1;
+* 'PSENLPModel': every element-Hessian approximation is updated with BFGS if the curvature condition holds, or withSR1 otherwise;
```
```suggestion
* 'PSENLPModel': every element-Hessian approximation is updated with BFGS if the curvature condition holds, or with SR1 otherwise;
```

dpo commented on docs/src/tutorial.md:
```diff
@@ -65,15 +70,27 @@ function jump_example(n::Int)
+* 'PCSNLPModel': each element-Hessian approximation with BFGS if it is classified as `convex`, or with SR1 otherwise;
+* 'PLBFGSNLP': every element-Hessian approximations is a LBFGS operator;
+* 'PLSR1NLPModel': every element-Hessian approximations is a LSR1 operator;
+* 'PLSENLPModel': by default, every element-Hessian approximations is a LBFGS operator as long as the curvature condition holds, otherwise it becomes a LSR1 operator.
```
Put model types between backticks instead of quotes.

dpo replied:
There is no PLCSNLPModel. Is that intentional? (I don't mean to add one to a list that is already long; I'm just wondering.)

paraynaud replied:
It is not intentionnal.
I implemented it for dense element matrices, I did not extended it to linear-operator element matrices.
I created an [issue](https://github.com/JuliaSmoothOptimizers/PartitionedStructures.jl/issues/66) in PartitionedStructures.jl for this addition.

dpo commented on docs/src/tutorial.md:
```diff
@@ -65,15 +70,27 @@ function jump_example(n::Int)
+* 'PBFGSNLPModel': every element-Hessian approximation is updated with BFGS;
+* 'PSR1NLPModel': every element-Hessian approximation is updated with SR1;
+* 'PSENLPModel': every element-Hessian approximation is updated with BFGS if the curvature condition holds, or withSR1 otherwise;
+* 'PCSNLPModel': each element-Hessian approximation with BFGS if it is classified as `convex`, or with SR1 otherwise;
```
```suggestion
* 'PCSNLPModel': each element-Hessian approximation with BFGS if it is classified as convex, or with SR1 otherwise;
```

dpo commented on docs/src/tutorial.md:
```diff
@@ -65,15 +70,27 @@ function jump_example(n::Int)
+* 'PSR1NLPModel': every element-Hessian approximation is updated with SR1;
+* 'PSENLPModel': every element-Hessian approximation is updated with BFGS if the curvature condition holds, or withSR1 otherwise;
+* 'PCSNLPModel': each element-Hessian approximation with BFGS if it is classified as `convex`, or with SR1 otherwise;
+* 'PLBFGSNLP': every element-Hessian approximations is a LBFGS operator;
```
```suggestion
* 'PLBFGSNLPModel': every element-Hessian approximations is a LBFGS operator;
```

dpo reviewed (commented):
If `AbstractPartiallySeparableNLPModel <: AbstractNLPModel`, you only need to implement `!` methods and `obj()`. The others are already implemented (unless you need to override some of them).

Such a large pull request is almost impossible to review though. It would take me days to go through it.

paraynaud commented:
> Such a large pull request is almost impossible to review though. It would take me days to go through it.

How should I proceed?
This time I had to rework entirely the repository.
Most of the code is copy-paste of the old code being reordered.

I mainly:
- defined the two abstract types;
- adapted the NLPModels methods depending these abstract types;
- defined all the partitioned NLPModels.jl (they are all very similar);
- implemented tests for them;
- deleted the old files.

The modifications impacted the documentation building and the tutorial content.

dpo commented:
I'll just trust you :-)

But as much as possible, we should aim for small focused pull requests (I recognize that in exceptional cases, it's not possible).

paraynaud commented:
@dpo I would like to merge this PR (it must be ready to go), to make a branch from the merge where I can integrate locally the PartitionedVectors and implement a TrunkSolver for all the PartiallySeparableNLPModels.

dpo commented on src/AbstractPNLPModels.jl:
```diff
@@ -0,0 +1,286 @@
+Evaluate `f(x)`, the objective function of `nlp` at `x`.
+"""
+function NLPModels.obj(
+  nlp::P,
+  x::AbstractVector{T},
+) where {T, S, P <: AbstractPartiallySeparableNLPModel{T, S}} 
```
```suggestion
  nlp::AbstractPartiallySeparableNLPModel{T, S},
  x::AbstractVector{T},
) where {T, S} 
```
You can make this simplification everywhere.

dpo reviewed (commented):
Ok for me after this minor change. The PR is too large to review.
